src/Lib.P2P/Routing/DistributedQuery.cs

```diff
@@ -0,0 +1,296 @@
+        {
+            log.Debug($"Q{Id} cancelled because DHT stopped.");
+            runningQuery.Cancel();
+        }
+
+        /// <summary>
+        ///   Ask the next peer the question.
+        /// </summary>
+        private async Task AskAsync(int taskId)
+        {
+            var pass = 0;
+            var waits = 20;
+            while (!runningQuery.IsCancellationRequested && waits > 0)
+            {
+                // Get the nearest peer that has not been visited.
+                var peer = Dht.RoutingTable
+                   .NearestPeers(QueryKey)
+                   .Where(p => !visited.ContainsKey(p))
+                   .FirstOrDefault();
+                if (peer == null)
+                {
+                    --waits;
+                    await Task.Delay(100);
+                    continue;
+                }
+
+                if (!visited.TryAdd(peer, peer)) continue;
+                ++pass;
+
+                // Ask the nearest peer.
+                await askCount.WaitAsync(runningQuery.Token).ConfigureAwait(false);
+                var start = DateTime.Now;
+                log.Debug($"Q{Id}.{taskId}.{pass} ask {peer}");
+                try
+                {
+                    using (var timeout = new CancellationTokenSource(askTime))
+                    using (var cts = CancellationTokenSource.CreateLinkedTokenSource(timeout.Token, runningQuery.Token))
+                    using (var stream =
+                        await Dht.SwarmService.DialAsync(peer, Dht.ToString(), cts.Token).ConfigureAwait(false))
+                    {
+                        // Send the KAD query and get a response.
+                        Serializer.SerializeWithLengthPrefix(stream, queryMessage, PrefixStyle.Base128);
+                        await stream.FlushAsync(cts.Token).ConfigureAwait(false);
+                        var response = await ProtoBufHelper.ReadMessageAsync<DhtMessage>(stream, cts.Token)
+                           .ConfigureAwait(false);
+
+                        // Process answer
+                        ProcessProviders(response.ProviderPeers);
+                        ProcessCloserPeers(response.CloserPeers);
+                    }
+
+                    var time = DateTime.Now - start;
+                    log.Debug($"Q{Id}.{taskId}.{pass} ok {peer} ({time.TotalMilliseconds} ms)");
+                }
+                catch (Exception e)
+                {
+                    Interlocked.Increment(ref failedConnects);
+                    var time = DateTime.Now - start;
+                    log.Warn($"Q{Id}.{taskId}.{pass} failed ({time.TotalMilliseconds} ms) - {e.Message}");
+
+                    // eat it
+                }
+                finally
+                {
+                    askCount.Release();
+                }
+            }
+        }
+
+        private void ProcessProviders(DhtPeerMessage[] providers)
+        {
+            if (providers == null)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)